### PR TITLE
CLOUDP-67537: add x509 type to dbusers create

### PR DESCRIPTION
**Pull request by robcarlan-mlab**
## Proposed changes

_Jira ticket:_ CLOUDP-67537

Closes #[issue number]

## Checklist

- [] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [] I have added tests that prove my fix is effective or that my feature works
- [] I have added any necessary documentation (if appropriate)
- [] I have run `make fmt` and formatted my code

## Further comments

x509 type:

```
./bin/mongocli atlas dbusers create --username managed --role backup --x509Type MANAGED
{
  "databaseName": "$external",
  "ldapAuthType": "NONE",
  "x509Type": "MANAGED",
  "awsIAMType": "NONE",
  "groupId": "5cac6a2ba6f239298a18a025",
  "roles": [
    {
      "roleName": "backup",
      "databaseName": "admin"
    }
  ],
  "username": "managed"
}

./bin/mongocli atlas dbusers create --username managed --role backup --x509Type MANAGED --password password
Error: cannot supply both x509 auth and password
```

Behaviour with none:
```
./bin/mongocli atlas dbusers create --username new --role backup --x509Type NONE
? Password: ***
{
  "databaseName": "admin",
  "ldapAuthType": "NONE",
  "x509Type": "NONE",
  "awsIAMType": "NONE",
  "groupId": "5cac6a2ba6f239298a18a025",
  "roles": [
    {
      "roleName": "backup",
      "databaseName": "admin"
    }
  ],
  "username": "new"
}

./bin/mongocli atlas dbusers create --username new --role backup
? Password:

./bin/mongocli atlas dbusers create --username new --role backup --x509Type NONE --password password
Error: POST https://cloud.mongodb.com/api/atlas/v1.0/groups/5cac6a2ba6f239298a18a025/databaseUsers: 409 (request "Conflict") A user with username new already exists.
...
```

### CODE REVIEW

File: internal/usage/usage.go [hunk 1]
```diff
@@ -46,6 +46,7 @@ const (
 	DiagnoseOut                     = "Optional output filename, if none given will use diagnose-archive.tar.gz."
 	LogStart                        = "Beginning of the period for which to retrieve logs."
 	LogEnd                          = "End of the period for which to retrieve logs."
+	X509Type                        = "X.509 method by which the provided username is authenticated. Valid values are NONE, MANAGED, or CUSTOMER."
```
Comment (gssbzn) [inline internal/usage/usage.go:49]: [nit-style] when we define valid values we tend to add them as a new line, you can see at line 141, we also move this multiline descriptions to the bottom as they mess up with the gofmt comand

File: internal/validate/validate.go [hunk 2]
```diff
@@ -98,3 +98,13 @@ func Credentials() error {
 	}
 	return nil
 }
+
+func EnsureFlagHasValidValue(value, flag string, validValues []string) error {
+	for _, v := range validValues {
```
Comment (gssbzn) [inline internal/validate/validate.go:103]: we have `search.StringInSlice` feel free to use it

File: internal/validate/validate.go [hunk 3]
```diff
@@ -98,3 +98,13 @@ func Credentials() error {
 	}
 	return nil
 }
+
+func EnsureFlagHasValidValue(value, flag string, validValues []string) error {
```
Comment (gssbzn) [inline internal/validate/validate.go:102]: [nit] on naming, we should read exported names along with the package name, `validate. EnsureFlagHasValidValue` it's a bit of a [stutter](https://stackoverflow.com/questions/45885985/ways-to-avoid-stuttering-in-go-package-and-struct-names), I think we can ditch the `Ensure` part

File: internal/cli/atlas/dbusers/create.go [hunk 4]
```diff
@@ -111,7 +153,7 @@ func CreateBuilder() *cobra.Command {
 	cmd.Flags().StringVarP(&opts.username, flag.Username, flag.UsernameShort, "", usage.Username)
 	cmd.Flags().StringVarP(&opts.password, flag.Password, flag.PasswordShort, "", usage.Password)
 	cmd.Flags().StringSliceVar(&opts.roles, flag.Role, []string{}, usage.Roles)
-	cmd.Flags().StringVar(&opts.authDB, flag.AuthDB, convert.AdminDB, usage.AuthDB)
+	cmd.Flags().StringVar(&opts.x509Type, flag.X509Type, "", usage.X509Type)
```
Comment (gssbzn) [inline internal/cli/atlas/dbusers/create.go:156]: is it safe if we set it to none here?
```suggestion
	cmd.Flags().StringVar(&opts.x509Type, flag.X509Type, X509TypeNone, usage.X509Type)
```

File: internal/cli/atlas/dbusers/create.go [hunk 5]
```diff
@@ -34,11 +35,23 @@ type CreateOpts struct {
...
-	authDB   string
+	x509Type string
 	roles    []string
 	store    store.DatabaseUserCreator
 }
 
+const (
+	X509TypeManaged  = "MANAGED"
+	X509TypeCustomer = "CUSTOMER"
+	X509TypeNone     = "NONE"
```
Comment (gssbzn) [inline internal/cli/atlas/dbusers/create.go:46]: [non-blocking] aws iam has also "NONE" this is more of a heads up

File: internal/usage/usage.go (hunk 1 above)
Comment (robcarlan-mlab) [reply internal/usage/usage.go:49]: Oh nice, I missed those ones at the bottom

File: e2e/atlas/dbusers_certs_test.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+		if err := json.Unmarshal(resp, &users); err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+		if len(users) == 0 {
+			t.Fatalf("expected len(users) > 0, got 0")
+		}
+	})
+
+	t.Run("Delete User", func(t *testing.T) {
+		cmd := exec.Command(cliPath, atlasEntity, dbusersEntity, "delete", username, "--force", "--authDB", "\\$external")
```
Comment (gssbzn) [inline e2e/atlas/dbusers_certs_test.go:104]: delete seems to be failing, I think us that double slash, we may need to urlencode `$external`(?)

Comment (robcarlan-mlab) [reply e2e/atlas/dbusers_certs_test.go]: answer was to not encode at all!

**Comment by robcarlan-mlab** [comment]
Tests seem to be working now, strangely my comment from yesterday didn't post

**Pull request merged**